state/state.go

```diff
@@ -211,12 +211,55 @@ func ControllerAccess(st *State, tag names.Tag) (permission.UserAccess, error) {
 	return st.UserAccess(tag.(names.UserTag), st.controllerTag)
 }
 
-// RemoveAllModelDocs removes all documents from multi-model
-// collections. The model should be put into a dying state before call
-// this method. Otherwise, there is a race condition in which collections
-// could be added to during or after the running of this method.
-func (st *State) RemoveAllModelDocs() error {
-	err := st.removeAllModelDocs(bson.D{{"life", Dead}})
+// setDyingModelToDead sets current dying model to dead.
+func (st *State) setDyingModelToDead() error {
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		model, err := st.Model()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if model.Life() != Dying {
+			return nil, errors.Trace(ErrModelNotDying)
+		}
+		ops := []txn.Op{{
+			C:      modelsC,
+			Id:     st.ModelUUID(),
+			Assert: isDyingDoc,
+			Update: bson.M{"$set": bson.M{
+				"life":          Dead,
+				"time-of-death": st.nowToTheSecond(),
+			}},
+		}, {
+			// Cleanup the owner:modelName unique key.
+			C:      usermodelnameC,
+			Id:     model.uniqueIndexID(),
+			Remove: true,
+		}}
+		return ops, nil
+	}
+	if err := st.db().Run(buildTxn); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// RemoveDyingModel sets current model to dead then removes all documents from
+// multi-model collections.
+func (st *State) RemoveDyingModel() error {
```

Comment:
We need a test for models in all life states - alive, dying, dead. Do we have all those tests or de we need to add one?

Reply:
yes, we have some tests for this, and I just added more tests for both RemoveDyingModel and setDyingModelToDead.